pybluetooth/pyusb_bt_sockets.py

```diff
@@ -0,0 +1,194 @@
+        except usb.core.USBError as e:
+            if e.errno == errno.ETIMEDOUT:
+                return None
+            else:
+                raise e
+
+        data = ''.join([chr(c) for c in data_array])  # Ugh.. array return val
+        data = "\4" + data  # Prepend H4 'Event' packet indicator
+        scapy_packet = HCI_Hdr(data)
+        LOG.debug("recv %s" % scapy_packet.lastlayer().summary())
+        LOG.debug("recv bytes: " + binascii.hexlify(data))
+        return scapy_packet
+
+    def send(self, scapy_packet):
+        data = str(scapy_packet)
+        LOG.debug("send %s" % scapy_packet.lastlayer().summary())
+        LOG.debug("send bytes: " + binascii.hexlify(data))
+        data = data[1:]  # Cut off the H4 'Command' packet indicator (0x02)
+        sent_len = self.pyusb_dev.ctrl_transfer(
+            data_or_wLength=data, **USB_HCI_CMD_REQUEST_PARAMS)
+        l = len(data)
+        if sent_len != l:
+            raise PyUSBBluetoothUserSocketException(
+                "Send failure. Sent %u instead of %u bytes" % (sent_len, l))
+
+
+def find_all_bt_adapters():
+    def bt_adapter_matcher(d):
+        # Check if the device is a "Single Function Primary Controller":
+        if (d.bDeviceClass == USB_DEVICE_CLASS_WIRELESS_CONTROLLER and
+            d.bDeviceSubClass == USB_DEVICE_SUB_CLASS_RF_CONTROLLER and
+            d.bDeviceProtocol == USB_DEVICE_PROTOCOL_BLUETOOTH):
+            return True
+
+        # Check if it's a composite device:
+        if not (d.bDeviceClass == USB_DEVICE_CLASS_MISCELLANEOUS and
+                d.bDeviceSubClass == USB_DEVICE_SUB_CLASS_COMMON_CLASS and
+                d.bDeviceProtocol == USB_DEVICE_PROTOCOL_IAD):
+            return False
+        for cfg in d:
+            bt_intf_descr = {
+                "bInterfaceClass": USB_DEVICE_CLASS_WIRELESS_CONTROLLER,
+                "bInterfaceSubClass": USB_DEVICE_SUB_CLASS_RF_CONTROLLER,
+                "bInterfaceProtocol": USB_DEVICE_PROTOCOL_BLUETOOTH,
+            }
+            intf = usb.util.find_descriptor(cfg, **bt_intf_descr)
+            if intf is not None:
+                return True
+        return False
+
+    devs = []
+    matchers = [CUSTOM_USB_DEVICE_MATCHER, bt_adapter_matcher]
+    for matcher in matchers:
+        if not matcher:
+            continue
+        devs += list(usb.core.find(find_all=True, custom_match=matcher))
+    return devs
+
+
+class PyUSBBluetoothNoAdapterFoundException(Exception):
+    pass
+
+
+def find_first_bt_adapter_pyusb_device_or_raise():
+    pyusb_devs = find_all_bt_adapters()
+    if len(pyusb_devs) == 0:
+        raise PyUSBBluetoothNoAdapterFoundException(
+            "No Bluetooth adapters found!")
+
+    def _is_usable_device(pyusb_dev):
+        try:
+            pyusb_dev.set_configuration()
+            PyUSBBluetoothHCISocket(pyusb_dev).hci_reset()
+            return True
+        except:
+            return False
+
+    pyusb_devs = filter(_is_usable_device, pyusb_devs)
```

Comment:
Added this, because it's possible that pyusb sees BT controllers that are already claimed, so try to send a HCI Reset as a means of verifying whether the device is usable or not.